google/resource_compute_attached_disk.go

```diff
@@ -0,0 +1,219 @@
+				Required:         true,
+				ForceNew:         true,
+				DiffSuppressFunc: compareSelfLinkOrResourceName,
+			},
+			"project": {
+				Type:     schema.TypeString,
+				ForceNew: true,
+				Computed: true,
+				Optional: true,
+			},
+			"zone": {
+				Type:     schema.TypeString,
+				ForceNew: true,
+				Computed: true,
+				Optional: true,
+			},
+			"device_name": {
+				Type:     schema.TypeString,
+				ForceNew: true,
+				Optional: true,
+				Computed: true,
+			},
+			"mode": {
+				Type:         schema.TypeString,
+				ForceNew:     true,
+				Optional:     true,
+				Default:      "READ_WRITE",
+				ValidateFunc: validation.StringInSlice([]string{"READ_ONLY", "READ_WRITE"}, false),
+			},
+		},
+	}
+}
+
+func resourceAttachedDiskCreate(d *schema.ResourceData, meta interface{}) error {
+	config := meta.(*Config)
+
+	zv, err := parseZonalFieldValue("instances", d.Get("instance").(string), "project", "zone", d, config, false)
+	if err != nil {
+		return err
+	}
+
+	instanceName := GetResourceNameFromSelfLink(d.Get("instance").(string))
+	diskName := GetResourceNameFromSelfLink(d.Get("disk").(string))
+
+	attachedDisk := compute.AttachedDisk{
+		Source:     fmt.Sprintf("projects/%s/zones/%s/disks/%s", zv.Project, zv.Zone, diskName),
+		Mode:       d.Get("mode").(string),
+		DeviceName: d.Get("device_name").(string),
+	}
+
+	op, err := config.clientCompute.Instances.AttachDisk(zv.Project, zv.Zone, instanceName, &attachedDisk).Do()
+	if err != nil {
+		return err
+	}
+
+	d.SetId(fmt.Sprintf("%s:%s", instanceName, diskName))
+
+	waitErr := computeSharedOperationWaitTime(config.clientCompute, op, zv.Project,
+		int(d.Timeout(schema.TimeoutCreate).Minutes()), "disk to attach")
+	if waitErr != nil {
+		d.SetId("")
+		return waitErr
+	}
+
+	return resourceAttachedDiskRead(d, meta)
+}
+
+func resourceAttachedDiskRead(d *schema.ResourceData, meta interface{}) error {
+	config := meta.(*Config)
+
+	zv, err := parseZonalFieldValue("instances", d.Get("instance").(string), "project", "zone", d, config, false)
+	if err != nil {
+		return err
+	}
+	d.Set("project", zv.Project)
+	d.Set("zone", zv.Zone)
+
+	instanceName := GetResourceNameFromSelfLink(d.Get("instance").(string))
+	diskName := GetResourceNameFromSelfLink(d.Get("disk").(string))
+
+	instance, err := config.clientCompute.Instances.Get(zv.Project, zv.Zone, instanceName).Do()
+	if err != nil {
+		return err
+	}
+
+	// Iterate through the instance's attached disks as this is the only way to
+	// confirm the disk is actually attached
+	ad := findDiskByName(instance.Disks, diskName)
+	if ad == nil {
+		log.Printf("[WARN] Referenced disk wasn't found attached to this compute instance. Removing from state.")
+		d.SetId("")
+		return nil
+	}
+
+	d.Set("device_name", ad.DeviceName)
+	d.Set("mode", ad.Mode)
+
+	// Force the referenced resources to a self-link in state because it's more specific then name.
+	instancePath, err := getRelativePath(instance.SelfLink)
```

Comment:
This is _probably_ fine. Some resources are annoying and insist that you give them a full URL instead of a relative one, so we generally use ConvertSelfLinkToV1 so we store the full URL. I doubt people are going to use this for interpolation so it doesn't really matter either way in this case, more of just a "hey letting you know that func exists too"